### PR TITLE
feat: report degraded constructs during markdown parsing

### PR DESCRIPTION
**Pull request by christianhg**
Schema-gating in `markdownToPortableText` degrades gracefully by design: an undeclared decorator drops its formatting, a table flattens, an image becomes its markdown source as plain text. Right for rendering pipelines, and invisible. Agent tooling patching rich text through this conversion needs the losses loud, an agent that wrote a table must learn the table didn't survive, and the accidental feedback channel it relied on (schema validation rejecting output converted against the wrong schema) disappears as soon as conversion runs against the correct schema.

One new option: `onDegradation`, a plain callback called at most once, after the walk, only when something degraded. Its report argument holds `degradations`, the encounter-ordered structured list (`{type, message, line?, snippet?}`, `type` a literal union to match on, `message` explicitly unstable, `snippet` the offending construct's text), and `message`, the canonical text written for the retry loop that reads it: each degradation states its effect in content terms, names the missing schema declaration as a constraint, quotes a snippet as the anchor (LLMs match strings far more reliably than they count lines), and identical degradations group into one line, forty undeclared-bold spans read as one rule. Enforcing is ordinary JavaScript rather than a library mode: `onDegradation: ({message}) => { throw new Error(message) }` propagates out of the conversion, pinned by a test. With the option unset, conversion degrades silently, which also removes the legacy `console.warn` on the eleven style-fallback paths: a library shouldn't log, visibility is opt-in.

A callback that never fires is a totality guarantee, so every site in the parser reports, including four the reviews found silent (structural-list paragraph styles, callout content styles, and the double-decline tiers on headings and flat blockquotes), the table-cell image demotion, and the task-checkbox strip at its consumption site so a structural path that preserves `checked` stays silent. Graceful block output is byte-identical to before, verified across default, sparse, and empty schemas, and pinned by the untouched suite.

`Degradation` is the only new export; its `type` field carries the literal union, which grows as new degradation sites report, so consumers compare against the values they handle rather than switching exhaustively. The option lives on the non-exported options type. The API went through four design iterations before this shape; the decision record with the rejected alternatives lives on the ticket.

---

> [!NOTE]
> **Medium Risk**
> Touches the core markdown-to-PT conversion path and changes observable logging (warn → silent by default); output shape is intended unchanged, but regression risk is moderate given breadth of degradation sites.
> 
> **Overview**
> Adds **`onDegradation`** to `markdownToPortableText`: when markdown can’t be represented in the schema (missing decorators, flattened tables, stripped task checkboxes, invalid `json:object` carriers, dropped object fields, and similar), conversion still returns the same lossy Portable Text as before, but can **report** every loss in one post-walk callback.
> 
> The report includes an encounter-ordered **`degradations`** array (exported **`Degradation`** type with stable **`type`**, optional **`line`** / **`snippet`**) and a grouped, line-sorted **`message`** for humans or agents. Callers can log, match on `type`, or **throw from the callback** to fail the import. Without the option, behavior stays silent—**`console.warn` on style-fallback paths is removed** in favor of opt-in reporting.
> 
> Implementation wires reporting through the markdown-it walk (new message catalog, snippet truncation, grouping rules) and extends default matchers to tag **fields dropped** when filtering to schema fields. Docs, README, changeset, and a large test suite cover edge cases (structural list/blockquote declines mirroring the flat path, table-cell images, callouts).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ff24b791512baf8953ac82665876c056bc312f4c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 1]
```diff
@@ -703,16 +807,23 @@ export function markdownToPortableText(
...
+
+        if (!blockquoteStyle) {
+          reportStyleFallback('blockquote', lineOf(token))
+        }
+
         const style =
-          consolidatedOptions.block.blockquote({
-            context: {schema: consolidatedOptions.schema},
-          }) ??
+          blockquoteStyle ??
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts:1154]: Fixed in 64aa2f333: the flat `blockquote_open` path now reports the second tier like the heading path does (`reportStyleFallback('normal', ...)` when the `normal` matcher also declines), pinned by a red-on-current test mirroring the heading double-decline pin. Good catch on the tier asymmetry.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 2]
```diff
@@ -274,6 +346,134 @@ function flattenTable(
...
+        if (b === undefined) {
+          return -1
+        }
+        return a - b
+      })
+
+    const suffix =
+      snippets.length === group.entries.length
+        ? `(${group.entries.length}\u00d7: ${snippets.map((snippet) => `"${snippet}"`).join(', ')})`
+        : `(lines ${linesAscending.join(', ')})`
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Triaged and hardened in the new head. The mechanism half of the finding is correct: the `(lines ...)` branch joined raw `line` values, so a multi-entry group with a line-less entry would print the literal word `undefined`. The cause half doesn't hold up: the flat-list and `addSpan` fallbacks have no token in scope (they're lazy block-openers), and probing every path we could construct through the public API (hr/html fallbacks, declining `block.normal` matchers, list items, blockquote content, callouts, table cells) always produced lined events — we could not manifest the `undefined` today. Fixed defensively anyway, since a line-less multi-entry group is one refactor away: undefined lines are now counted but never printed (`(3×: lines 1, 5)` when partially lined, `(2×)` when unlined). No public-API pin exists because no reachable input produces the shape; the guard is exercised by the existing grouped-message tests.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 3]
```diff
@@ -1388,6 +1691,13 @@ export function markdownToPortableText(
 
           if (blockImageObject) {
             if (inTableCell) {
+              report({
+                type: 'image-block-to-inline',
+                message:
+                  "Embedded image inline: table cell can't hold a block-level `image`",
+                line: inlineLine(),
+              })
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Confirmed and fixed in the new head. The event fired at demotion time, before `td_close`'s sole-image lift undid the demotion, so `| alt |` with the default schema reported a degradation on a lossless conversion (your repro, verified: the cell ends up holding the block-level `image`). The report now defers to `td_close`: demoted images are marked during the cell walk, and only those the lift doesn't consume report. The genuinely degrading shape (image declared only as a block object, mixed cell content, so the image survives as an inline child) still fires exactly once, pinned by a contrast test. One correction to the finding's premise: the default path no longer `console.warn`s at all, the default has been silent since an earlier revision; only an `onDegradation` consumer would have seen the false positive.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 4]
```diff
@@ -479,6 +828,19 @@ export function markdownToPortableText(
...
+    // to output needing that style; a callout holding nothing but, say, a
+    // standalone image discards its placeholder block without ever
+    // reaching here (see the `inline` case's standalone-image branch), and
+    // never degraded.
+    if (calloutPendingStyleFallbacks.length > 0) {
+      for (const name of calloutPendingStyleFallbacks) {
+        reportStyleFallback(name, calloutStartLine)
+      }
+      calloutPendingStyleFallbacks = []
+    }
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Confirmed and fixed in the new head: the flush now gates on the committing block actually carrying the fallen-back style, so a heading inside a callout (which keeps its own `h1`) no longer reports. Pinned red-on-current, with a contrast pin proving a plain paragraph in the same callout still fires (that block does take the declined style).

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 5]
```diff
@@ -1693,13 +2404,22 @@ export function markdownToPortableText(
...
+                })
                 addSpan(`![${alt}](${src})`)
                 break
               }
 
               // Block image supported - flush current block and add as block-level
               // Skip if we're in a table cell (images in cells are handled differently)
               if (inTableCell) {
+                demotedTableImages.add(blockImageObject as PortableTextObject)
+
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Confirmed and fixed in the new head: the mixed-content cell demotion site now runs the same `reportFieldsDropped` check as every other accepted-object site, so a schema-omitted `alt`/`title` reports `fields-dropped` alongside `image-block-to-inline`. Pinned red-on-current with both events asserted, the fields message naming `alt`.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 6]
```diff
@@ -479,6 +852,25 @@ export function markdownToPortableText(
...
+    // never degraded.
+    if (
+      calloutPendingStyleFallbacks.length > 0 &&
+      currentBlock.style === currentBlockquoteStyle
+    ) {
+      for (const name of calloutPendingStyleFallbacks) {
+        reportStyleFallback(name, calloutStartLine)
+      }
+      calloutPendingStyleFallbacks = []
+    }
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Confirmed and fixed in the new head, by provenance instead of value: blocks are now marked at the moment they actually receive the callout's fallen-back content style, and the flush fires only for marked blocks. A heading that independently falls back to `normal` no longer collides with a callout style that also fell to `normal`. Pinned: that exact input emits only the `h1` event, with contrast pins keeping the true-positive paragraph case firing.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 7]
```diff
@@ -864,47 +1317,97 @@ export function markdownToPortableText(
...
+                  continue
+                }
+
+                mergeTarget = {
+                  ...textBlock,
+                  listItem: itemListType as string,
+                  level,
+                  ...(itemChecked === undefined ? {} : {checked: itemChecked}),
+                }
+                pushBlock(mergeTarget)
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Real, fixed: the decline merge now requires provenance (a block the flat path would treat as a plain paragraph) plus matching `style` before merging, restoring the flat-mirror invariant. Pinned by canonical flat-vs-declined equality for both the nested-blockquote and degraded-fence-in-item cases.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 8]
```diff
@@ -1342,6 +1929,21 @@ export function markdownToPortableText(
...
+        // child of the cell's text block: report it now that the verdict is
+        // known, instead of at demotion time.
+        for (const demotedImage of demotedInlineImages) {
+          if (demotedImage !== liftedImage) {
+            report({
+              type: 'image-block-to-inline',
+              message:
+                "Embedded image inline: table cell can't hold a block-level `image`",
+              line: currentTable?.line,
+            })
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Fixed: the deferred report now attaches the same `alt`-or-`src` snippet the non-deferred `image-block-to-inline` sites use, pinned in both mixed-cell tests.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 9]
```diff
@@ -274,6 +350,171 @@ function flattenTable(
...
+
+    const count = group.entries.length
+    const suffix =
+      snippets.length === count
+        ? `(${count}\u00d7: ${capList(snippets.map((snippet) => `"${snippet}"`))})`
+        : linesAscending.length > 0
+          ? `(${count}\u00d7: lines ${capList(linesAscending.map(String))})`
+          : `(${count}\u00d7)`
+
+    return `- ${group.base} ${suffix}`
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Fixed: line lists dedupe before capping, so same-line groups render `(3×: lines 4)`; the count carries the multiplicity. Pinned.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 10]
```diff
@@ -750,21 +1144,40 @@ export function markdownToPortableText(
             if (blockquoteObject) {
               pushBlock(blockquoteObject)
             } else {
-              // Matcher returned undefined: fall back to flat-style by
-              // re-emitting each content block with `style: 'blockquote'`.
-              const blockquoteStyle =
-                consolidatedOptions.block.blockquote({
-                  context: {schema: consolidatedOptions.schema},
-                }) ??
+              // Matcher returned undefined: fall back to exactly what the
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Not a bug, verified empirically: a schema-declared `blockquote`/`list` object missing its content-carrying field doesn't destroy the payload silently. The conversion takes the flat path (content survives as flat blocks/list items) and the flat path's own events fire (`style-fallback` / `list-flattened` in the probes). Accepted container objects only arise from consumer-provided `types.*` matchers, whose output the library trusts by design: `fields-dropped` reports default-matcher field filtering only.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 11]
```diff
@@ -750,21 +1188,40 @@ export function markdownToPortableText(
...
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Not a bug, and the claimed mechanism doesn't exist: there is no default matcher for `blockquote`/`list` container objects (`defaultOptions.types` covers `code`, `horizontalRule`, `html`, `image`, `callout`, `table` only), so no code path tags a consumer-returned container with dropped fields. A consumer `types.blockquote` result passes through untouched: probed with the documented `({value}) => ...` signature, `content` and extra fields both survive intact, zero events. `fields-dropped` reports default-matcher field filtering by design; consumer-built objects are the consumer's own output.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 12]
```diff
@@ -750,21 +1188,40 @@ export function markdownToPortableText(
...
               for (const block of contentBlocks) {
-                if (block._type === 'block') {
+                if (
+                  block._type === 'block' &&
+                  !/^h[1-6]$/.test((block as PortableTextTextBlock).style ?? '')
+                ) {
                   pushBlock({
                     ...(block as PortableTextTextBlock),
-                    style: blockquoteStyle,
+                    style: fallbackStyle,
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Fixed: the decline restyle rebuilt blocks by spread, losing `plainParagraphBlocks` WeakSet membership, so an enclosing list decline failed the stampable gate and the block left the list. The clone now re-enters the set, and the flat-mirror equality for a declined blockquote inside a declined list is pinned red-on-current.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 13]
```diff
@@ -1435,17 +2107,29 @@ export function markdownToPortableText(
...
 
           if (inlineImageObject) {
+            reportFieldsDropped(inlineImageObject, inlineLine())
+            report({
+              type: 'image-block-to-inline',
+              message:
+                degradationMessage['image-block-to-inline']('no-block-image'),
+              line: inlineLine(),
+              snippet: truncateSnippet(alt || src),
+            })
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Real, fixed: the inline-matched branch reported immediately instead of deferring through the `td_close` sole-image-lift verdict like the block-matched branch. `demotedTableImages` is now a `WeakMap` carrying the cause, and both branches defer. Pinned: sole cell image under an inline-only schema converts losslessly with zero events; mixed-cell contrasts stay pinned.

File: packages/markdown/src/to-portable-text/markdown-to-portable-text.ts [hunk 14]
```diff
@@ -750,21 +1144,40 @@ export function markdownToPortableText(
...
                   context: {schema: consolidatedOptions.schema},
-                }) ??
-                'blockquote'
+                })
+
+              if (!resolvedStyle) {
+                reportStyleFallback('normal', frame.line)
+              }
+
+              const fallbackStyle = resolvedStyle ?? 'blockquote'
```
[the remark this answers is not in the document]
Comment (christianhg) [reply packages/markdown/src/to-portable-text/markdown-to-portable-text.ts]: Couldn't reproduce against the current head, in four probe variants (`types.blockquote` declining with and without a schema `blockquote` style, and with consumer `styles.blockquote`/`styles.normal` matchers declining): declined output is byte-identical to the flat path and the events match exactly, no `normal`-vs-`blockquote` contradiction. This comment targeted the head before the decline path was rewritten to mirror the flat path by provenance; the newer run against that head didn't restate it. If you can still trigger a contradicting pair, a concrete input + schema would let me pin it.
